packages/client-engine-runtime/src/interpreter/DataMapper.ts

```diff
@@ -0,0 +1,166 @@
+import { PrismaValueType, ResultNode } from '../QueryPlan'
+import { assertNever } from '../utils'
+import { PrismaObject, Value } from './scope'
+
+export function applyDataMap(data: Value, structure: ResultNode): Value {
+  switch (structure.type) {
+    case 'object':
+      return mapArrayOrObject(data, structure.fields)
+
+    case 'value':
+      return mapValue(data, structure.resultType)
+
+    default:
+      // @ts-ignore
+      assertNever(structure, `Invalid data mapping type: '${structure.type}'`)
+  }
+}
+
+function mapArrayOrObject(data: Value, fields: Record<string, ResultNode>): PrismaObject | PrismaObject[] | null {
+  if (data === null) return null
+
+  if (Array.isArray(data)) {
+    const rows = data as PrismaObject[]
+    return rows.map((row) => mapObject(row, fields))
+  }
+
+  if (typeof data === 'object') {
+    const row = data as PrismaObject
+    return mapObject(row, fields)
+  }
+
+  throw new Error(`DataMapper[1]: Expected an array or an object, got: ${typeof data}`)
+}
+
+// Recursive
+function mapObject(data: PrismaObject, fields: Record<string, ResultNode>): PrismaObject {
+  if (typeof data !== 'object') {
+    throw new Error(`DataMapper[2]: Expected an object, but got '${typeof data}'`)
+  }
+
+  const result = {}
+  for (const [name, node] of Object.entries(fields)) {
+    switch (node.type) {
+      case 'object':
+        if (name in data) {
+          result[name] = mapArrayOrObject(data[name], node.fields)
+        } else {
+          throw new Error(
+            `DataMapper[3]: Missing data field: '${name}'; ` +
+              `node: ${JSON.stringify(node)}; data: ${JSON.stringify(data)}`,
+          )
+        }
+        break
+
+      case 'value':
+        if (node.dbName in data) {
+          result[name] = mapValue(data[node.dbName], node.resultType)
+        } else {
+          throw new Error(
+            `DataMapper[4]: Missing data field: '${node.dbName}'; ` +
+              `node: ${JSON.stringify(node)}; data: ${JSON.stringify(data)}`,
+          )
+        }
+        break
+
+      default:
+        // @ts-ignore
+        assertNever(node, `Invalid data mapping node type: '${node.type}'`)
+    }
+  }
+  return result
+}
+
+function mapValue(value: unknown, resultType: PrismaValueType): unknown {
+  if (value === null) return null
+
+  if (typeof resultType === 'string') {
```

Comment:
We may want to choose a different encoding (e.g. `#[serde(tag = "type")]`) to make it easier to work with this enum in TypeScript.